BlueprintUI/Sources/AttributedText/AttributedText.swift

```diff
@@ -0,0 +1,149 @@
+/// text[range].color = .blue
+///
+/// // Render the attributed text
+/// let label = AttributedLabel(attributedText: text.attributedString)
+/// ```
+///
+@dynamicMemberLookup public struct AttributedText {
+
+    /// The wrapped string, with no attributes.
+    public let string: String
+
+    /// An `NSAttributedString` representation of the attributed text.
+    public var attributedString: NSAttributedString {
+        mutableAttributedString
+    }
+
+    private let mutableAttributedString: NSMutableAttributedString
+
+    /// Create some `AttributedText` from a plain string.
+    public init(string: String) {
+        self.string = string
+        self.mutableAttributedString = NSMutableAttributedString(string: string)
+    }
+
+    /// Create some `AttributedText` from an attributed string. The attributes are preserved.
+    public init(attributedString: NSAttributedString) {
+        self.string = attributedString.string
+        self.mutableAttributedString = NSMutableAttributedString(attributedString: attributedString)
+    }
+
+    public func range<StringType: StringProtocol>(of aString: StringType) -> Range<String.Index>? {
+        string.range(of: aString)
+    }
+
+    /// Dynamic member getter or setter for any attributes defined on `TextAttributeContainer`.
+    /// Applies the attribute to the entire range of text, for example:
+    ///
+    /// ```swift
+    /// var text = AttributedText(string: "Hello, world")
+    /// text.font = .systemFont(ofSize: 20)
+    /// ```
+    /// Note that only attributes applying to the entire range will be returned. For example, if the text
+    /// has two different `font` attributes, then `text.font` will be `nil`.
+    ///
+    public subscript<Value>(dynamicMember keyPath: WritableKeyPath<TextAttributeContainer, Value>) -> Value {
+        get {
+            self[entireRange][keyPath: keyPath]
+        }
+        set {
+            self[entireRange][keyPath: keyPath] = newValue
+        }
+    }
+
+    /// Get or set a `TextAttributeContainer` for the provided range of text. This allows you to set attributes
+    /// for specific ranges using strong types:
+    ///
+    /// ```swift
+    /// var text = AttributedText(string: "Hello, world")
+    /// let range = text.string.range(of: "Hello")!
+    /// text[range].font = .systemFont(ofSize: 20)
+    /// ```
+    ///
+    /// Note that the returned `TextAttributeContainer` will only contain attributes that apply to the entire subscript
+    /// range. (Setting an attribute will set it across the subscript range regardless of any existing contents).
+    ///
+    public subscript(range: Range<String.Index>) -> TextAttributeContainer {
+        get {
+            let range = NSRange(range, in: string)
+            return makeAttributeStore(range: range)
+        }
+        set {
+            let range = NSRange(range, in: string)
+            mutableAttributedString.addAttributes(newValue.storage, range: range)
```

Comment:
`NSMutableAttributedString` is a mutable reference (class) type. This struct is a class type; you need to implement copy on write semantics here otherwise each copy of the struct is gonna reference the same underlying mutable string

Reply:
Ah yep, very good catch...

Reply:
Added a test for this and pushed up a fix

Reply:
Nice